### PR TITLE
[PVR][addons] PVR Add-on API changes + Implementation: Custom Timer Settings

### PR DESCRIPTION
**Pull request by ksooo**
This is something I had on my todo list for quite a while already: Add-on supplied custom timer settings.

With this PR PVR client add-ons can define custom string and integer settings for timers. Definition includes a localized setting name, possible setting values (incl. localized description per value), default value and other attributes.

The definitions are attached to timer types. If present, the settings will be integrated automatically in the timer settings dialog and appear from users perspective as every other timer setting. Values can be changed there and will be handed back to the add-on.

Kodi will (and can) not interpret the settings in any way. The data will just be routed form the add-on to Kodi and vice versa.

Lots of code was added for both API and core for the implementation of this feature. However, For an add-on it should be quite simple and not much code is needed. A will open a PR with a reference implementation for pvr.hts soon.

As some new API structs and enums are introduced, the needed boilerplate code for the C++ API wrapper summed up to > 1K LOC. For the core code I wanted to have it in a real OOP style, so couple of new classes here, which should ensure readability and long term maintainability of the new code.

Last, the concept is so designed, that the API can easily be extended to add other setting types (currently "only" string and integer settings are supported). Strict encapsulation of the core implementation also helps here. To proof this, I started with integer settings and added string settings later. ;-)

Best, although we face an API change here, no add-on must be recompiled. I added two runtime API version checks were needed. That's all that was needed. One more time: well structured code helps maintainability.

I runtime-tested the changes carefully, including setups were "old" add-ons were used together with a "new" Kodi. I found no problems (does no mean that there are no bugs, though).

@phunkyfish I hope that you find some time to review the changes soon. I know it is a lot of code, but I hope that the design I've chosen and the clean implementation helps here.

### CODE REVIEW

File: xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Timers.h [hunk 1]
```diff
@@ -822,7 +866,7 @@ class PVRTimerType : public DynamicCStructHdl<PVRTimerType, PVR_TIMER_TYPE>
   /// @brief **optional**\n
   /// Array containing the possible values of @ref PVRTimer::SetMaxRecordings().
   ///
-  /// @param[in] maxRecordings List of lifetimes values
+  /// @param[in] maxRecordings List of max recordings values
```
Comment (phunkyfish) [inline xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Timers.h:869]: I assume this was just an error in the docs?

Comment (phunkyfish) [inline xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Timers.h:869]: I assume this was just an error in the docs?

Comment (ksooo) [reply xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Timers.h:869]: Yep, c/p error I guess

**Comment by ksooo** [comment]
Some more information... The concept in a nutshell:

* If an add-on wants to use timer properties that shall be manageable using Kodis timer settings dialog, but are not pre-defined by Kodi PVR core, it can define settings itself now, using an array of `PVR_SETTING_DFINITION` (an id, a name, possible values, default value, ….).
* The definitions are attached to the add-on defined timer types, so every type can have own custom settings.

* Timers itself can now carry a value per custom add-on defined custom setting, in an array of `PVR_SETTING_KEY_VALUE_PAIR` (a ref to the setting via id, the value).

* Kodi core picks up the definitions once the timer types come in and wraps them into a vector of `CPVRTimerSettingDefinition` per type.

* Once a timer comes in, it’s custom properties will be stored in its `CustomPropsMap` member.

* Timer settings dialog reads the definitions and incorporates them into the list of settings it presents to the user. From user’s perspective, the custom settings are in no way different from the settings pre-defined by Kodi PVR core.

* If the timer setting dialog gets closed, timer’s (changed) custom properties will be written back to timer’s custom props map.

* Then, the timer changes will be transferred back to the client, like usual, using `UpdateTimer` API call.

* The client add-on then can pick up the (changed) custom property value(s) and do whatever it wants with it.

Here is an example how it looks like for a pvr.hts defined timer rule containing three custom settings:

image

File: xbmc/pvr/addons/PVRClient.cpp [hunk 2]
```diff
@@ -263,6 +282,7 @@ class CAddonTimer : public PVR_TIMER
   const std::string m_directory;
   const std::string m_summary;
   const std::string m_seriesLink;
+  std::unique_ptr<PVR_SETTING_KEY_VALUE_PAIR[]> m_customProps;
```
Comment (phunkyfish) [inline xbmc/pvr/addons/PVRClient.cpp:285]: Why is this an array as opposed to being a vector? We're in C++ here, so using an array feels wrong

If it's just to make conversion to/from the C API, then fine.

**Review by phunkyfish — COMMENTED** [review]
The design is solid. Just a comment so far.

File: xbmc/pvr/addons/PVRClient.cpp (hunk 2 above)
Comment (ksooo) [reply xbmc/pvr/addons/PVRClient.cpp:285]: > If it's just to make conversion to/from the C API, then fine.

That's is exactly the purpose. Translate C++ to C data structures, to be able to call the C API.

Comment (phunkyfish) [reply xbmc/pvr/addons/PVRClient.cpp:285]: Ok, all good.

File: xbmc/pvr/settings/PVRStringSettingValues.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  : m_defaultValue(defaultValue)
+{
+  if (values && valuesSize > 0)
+  {
+    m_values.reserve(valuesSize);
+    for (unsigned int i = 0; i < valuesSize; ++i)
+    {
+      const std::string value{values[i].strValue ? values[i].strValue : ""};
+      const char* desc{values[i].strDescription};
+      std::string strDescr{desc ? desc : ""};
```
Comment (phunkyfish) [inline xbmc/pvr/settings/PVRStringSettingValues.cpp:32]: `strDesc` would be a more commonplace abbreviation. The current var name made me look twice ;)

**Review by phunkyfish — APPROVED** [review]
Just one very minor, minor. But not a blocker.

I even ran this with a contrived example to see how easy it is from the add-on side and it's super lightweight. Super clean PR, nice work!

File: xbmc/pvr/settings/PVRStringSettingValues.cpp (hunk 3 above)
Comment (ksooo) [reply xbmc/pvr/settings/PVRStringSettingValues.cpp:32]: O yes, seems more like a typo + autocomplete afterwards to me. Will change.

**Review by phunkyfish — APPROVED** [review]

**Pull request merged**